### DISCUSSION TITLE
Why are `Sparta::Buffer::erase(const iter&)` return type `void`?  Does `Sparta::Buffer:erase(const iter&)` invalidate iterators? How to handle `erase()` in an iterator loop?

### DISCUSSION THREAD

**Discussion by timsnyder** [sparcians/map#459] [Q&A] [answered]
In https://github.com/sparcians/map/blob/e876f588ec53eab68c24e4f70adf47c3c7e901c2/sparta/sparta/resources/Buffer.hpp#L663-L683

The return types of `Sparta::Buffer::erase(const iter&)`  are `void`.  Based on the STL semantics for `erase()`, I would expect them to return a copy of the iterator adjusted for the erasure.  Does `Sparta::Buffer::erase()` invalidate the iterator?

For example, with a `std::vector`, I would expect to need to handle iterator invalidation something like:
```
#include <vector>


int main() {
    std::vector<int> junk = {1,3,5,7,9};

    for(auto itr = junk.begin(); itr != junk.end();) {
        if (*itr == 5) {
            itr = junk.erase(itr);
        } else {
            itr++;
        }
    }
}
```

Do I need to worry about the same type of thing with `Sparta::Buffer` or is the iterator somehow not invalidated by `Sparta::Buffer::erase()`?

**Comment by klingaard** [comment, marked as answer]
No, there's no good reason that it doesn't return the next iterator.  I think this was based on an older C++ STL design.  In fact, there are a few issues with Buffer, Queue, etc that I already started documenting here: https://github.com/sparcians/map/issues/247

Feel free to convert this to an issue and either make the changes or leave it unassigned and I'll grab it.

**Comment by klingaard** [comment]
Converted to issue #464
